### ISSUE TITLE
AdaptiveSampler#sampled? works in the unexpected way

### ISSUE THREAD

**Issue by romul** [newrelic/newrelic-ruby-agent#869]
I guess AdaptiveSampler#sampled? works in the unexpected way

## Description
I expected that sampling algorithm would take random examples of desired transaction, but the code looks like it just takes the first ones (target_samples+1 count) in each period (period_duration).

## Expected Behavior
I expected to have 10% of transaction occurrences in the period to be sampled in the distributed tracing.

## Additional context
Let's take a look at the code:
https://github.com/newrelic/newrelic-ruby-agent/blob/8.2.0/lib/new_relic/agent/adaptive_sampler.rb#L24..L40

It seems it is overcomplicated, b/c atm it works in fact almost like that:
```
          sampled = if @first_period
            @sampled_count < 10
          elsif @sampled_count <= @target
            true # b/c @seen_last = elapsed_periods > 1 ? 0 : @seen, so it equals 0 unless @first_period
          else
             false 
          end
 ```
 
The condition `rand(@seen) < (@target ** (@target / @sampled_count) - @target ** 0.5)` is a nonsense, b/c `@target` & `@sampled_count` are both Integer, so if `@sampled_count` would be greater than `@target`, it gives 
`(@target ** 0 - @target ** 0.5)`, which is negative and always gives `false` as the result of comparison. 

As a result it's a very strange sampling algorithm, it takes the first `@target - 1` (9 by default) transactions and than randomize to take another two from 10th, 11th, 12th or 13th at maximum.

**Comment by kaylareopelle** [comment]
Hi @romul, thank you for reaching out with your concern about the way `AdaptiveSampler#sampled?` behaves! 

You're correct, the integer division is causing the value returned by `rand(@seen) < (@target ** (@target / @sampled_count) - @target ** 0.5)` to always be zero. Thank you for catching this and breaking down the problem for us.

Though replacing integers with floats when dividing is a solution to this problem, we need to talk with our team about the implications of changing the existing algorithm before moving forward.

**Comment by kaylareopelle** [comment]
Hi @romul  - I wanted to let you know a change in response to your issue is included in the newly published version 8.3.0 of the agent. Thanks for your contribution!

**Comment by romul** [comment]
Thank you @kaylareopelle 👍
